src/include/self_driving/model_server/model_server_manager.h

```diff
@@ -0,0 +1,239 @@
+ */
+static constexpr const unsigned char MODEL_SERVER_ERROR_BINARY = 128;
+
+/**
+ * ModelServerFuture is a wrapper over a condition_variable, and the condition.
+ *
+ * It is used to build synchronous API over asynchronous function calls:
+ * ```c++
+ *      ModelServerFuture future;
+ *      AsyncCall(future);
+ *
+ *      auto result = future.Wait();
+ *      bool success = result.first;
+ *      auto data = result.second;
+ * ```
+ *
+ * @tparam Result the future's result
+ */
+template <class Result>
+class ModelServerFuture {
+ public:
+  /**
+   * Initialize the future with a given mutex
+   * @param mtx
+   */
+  ModelServerFuture() = default;
+
+  /**
+   * Suspends the current thread and wait for the result to be ready
+   * @return Result, and success/fail
+   */
+  std::pair<Result, bool> Wait() {
+    {
+      std::unique_lock<std::mutex> lock(mtx_);
+
+      // Wait for it to be done
+      cvar_.wait(lock, [&] { return done_.load(); });
+    }
+
+    return {result_, success_};
+  }
+
+  /**
+   * Indicate a future is done
+   * @param result
+   */
+  void Done(Result result) {
+    {
+      std::unique_lock<std::mutex> lock(mtx_);
+      result_ = result;
+      done_ = true;
+      success_ = true;
+    }
+    cvar_.notify_all();
+  }
+
+  /**
+   * Indicate the future could not complete due to failure of the Async call
+   */
+  void Fail(const std::string &reason) {
+    {
+      std::unique_lock<std::mutex> lock(mtx_);
+      done_ = true;
+      success_ = false;
+      fail_msg_ = reason;
+    }
+    cvar_.notify_all();
+  }
+
+  /**
+   * Return why the operation has failed
+   */
+  const std::string &FailMessage() const { return fail_msg_; }
+
+ private:
+  /** Result for the future **/
+  Result result_;
+
+  /** Condition variable **/
+  std::condition_variable cvar_;
+
+  /** If async operation done */
+  std::atomic<bool> done_ = false;
+
+  /** If async operation succeeds */
+  std::atomic<bool> success_ = false;
+
+  /** Reason for failure **/
+  std::string fail_msg_;
+
+  /** Mutex associated with the condition variable **/
+  std::mutex mtx_;
+};
+
+/**
+ * Interface for ModelServerManager related operations
+ */
+class ModelServerManager {
+  enum class Callback : uint64_t { NOOP = 0, CONNECTED, DEFAULT };
+
+ public:
+  /**
+   * Construct a ModelServerManager with the given executable script to the python ModelServer
+   * @param model_bin python script path
+   * @param messenger Messenger pointer
+   */
+  ModelServerManager(const std::string &model_bin, const common::ManagedPointer<messenger::Messenger> &messenger);
+
+  /**
+   * Stop the python ModelServer when exits
+   */
+  ~ModelServerManager() { StopModelServer(); }
+
+  /**
+   * Stop the python-ModelServer daemon by sending a message to the Python model server
+   */
+  void StopModelServer();
+
+  /**
+   * Check if the model server has started.
+   * The user of this function should poll this until the pid changes.
```

Comment:
Wait, why does the user also need to poll the PID? Shouldn't the PID only be changed if connected_ is changed as well?